File: lib/notification_service/discord/formatter.ex
defmodule NotificationService.Discord.FormatterBehaviour do
  @moduledoc """
  Behaviour for Discord message formatting.
  """

  @callback format(routing_key :: String.t(), payload :: map()) :: map()
end

defmodule NotificationService.Discord.Formatter do
  @moduledoc """
  Transforms message payloads into Discord embed maps based on routing key category.
  """

  @behaviour NotificationService.Discord.FormatterBehaviour

  @color_green 0x00FF00
  @color_red 0xFF0000
  @color_yellow 0xFFFF00
  @color_purple 0x800080
  @color_orange 0xFF8C00
  @color_blue 0x0099FF

  @impl true
  @spec format(String.t(), map()) :: map()
  def format("ci.success", payload), do: format_ci(payload, "success", @color_green)
  def format("ci.failure", payload), do: format_ci(payload, "failure", @color_red)

  def format("pr.opened", payload), do: format_pr(payload, "opened", @color_yellow)
  def format("pr.merged", payload), do: format_pr(payload, "merged", @color_purple)
  def format("pr.closed", payload), do: format_pr(payload, "closed", @color_red)

  def format("downloads.complete", payload),
    do: format_download(payload, "Complete", @color_green)

  def format("downloads.removed", payload), do: format_download(payload, "Removed", @color_orange)

  def format("deploy.success", payload), do: format_deploy(payload, "success", @color_green)
  def format("deploy.failure", payload), do: format_deploy(payload, "failure", @color_red)

  def format("polling.failure", payload), do: format_polling_failure(payload)

  def format("dlq." <> _service, payload), do: format_dlq(payload)

  def format(_routing_key, payload), do: format_unknown(payload)

  defp format_ci(payload, conclusion, color) do
    repo = Map.get(payload, "repository", "unknown")
    run_url = Map.get(payload, "run_url")

    %{
      title: "CI: #{repo}",
      color: color,
      url: run_url,
      fields: [
        %{name: "Status", value: Map.get(payload, "status", "unknown"), inline: true},
        %{name: "Result", value: conclusion, inline: true}
      ],
      timestamp: DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  defp format_pr(payload, action, color) do
    pr_number = Map.get(payload, "pr_number", "?")
    pr_title = Map.get(payload, "pr_title", "Untitled")
    author = Map.get(payload, "author", "unknown")
    pr_url = Map.get(payload, "pr_url")

    %{
      title: "PR ##{pr_number}: #{pr_title}",
      description: "#{action} by @#{author}",
      color: color,
      url: pr_url,
      timestamp: DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  defp format_download(payload, action, color) do
    name = Map.get(payload, "name", "Unknown")
    size = Map.get(payload, "size")
    category = Map.get(payload, "category")
    save_path = Map.get(payload, "savePath")

    fields =
      [
        if(size, do: %{name: "Size", value: format_bytes(size), inline: true}),
        if(category, do: %{name: "Category", value: category, inline: true}),
        if(save_path, do: %{name: "Path", value: save_path, inline: false})
      ]
      |> Enum.reject(&is_nil/1)

    %{
      title: "Download #{action}: #{name}",
      color: color,
      fields: fields,
      timestamp: DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  defp format_deploy(payload, conclusion, color) do
    repo = Map.get(payload, "repository", "unknown")

    fields =
      [
        %{name: "Status", value: conclusion, inline: true},
        if(Map.get(payload, "duration"),
          do: %{name: "Duration", value: Map.get(payload, "duration"), inline: true}
        )
      ]
      |> Enum.reject(&is_nil/1)

    %{
      title: "Deploy: #{repo}",
      color: color,
      fields: fields,
      timestamp: DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  defp format_polling_failure(payload) do
    service = Map.get(payload, "service", "unknown")
    error = Map.get(payload, "error", "Unknown error")

    %{
      title: "Polling Failure: #{service}",
      color: @color_red,
      fields: [
        %{name: "Error", value: error, inline: false}
      ],
      timestamp: DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  defp format_dlq(payload) do
    service = Map.get(payload, "service", "unknown")
    queue = Map.get(payload, "queue", "unknown")
    error = Map.get(payload, "error", "Unknown error")
    retry_count = Map.get(payload, "retryCount", 0)

    %{
      title: "DLQ Alert: #{service}",
      description: "#{queue} queue exceeded retry limit",
      color: @color_red,
      fields: [
        %{name: "Error", value: error, inline: false},
        %{name: "Retries", value: to_string(retry_count), inline: true},
        %{name: "Queue", value: queue, inline: true}
      ],
      timestamp: DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  defp format_unknown(payload) do
    description =
      case Jason.encode(payload, pretty: true) do
        {:ok, json} -> "```json\n#{String.slice(json, 0, 1900)}\n```"
        _ -> inspect(payload)
      end

    %{
      title: "Notification",
      description: description,
      color: @color_blue,
      timestamp: DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  @spec format_bytes(number()) :: String.t()
  defp format_bytes(bytes) when is_number(bytes) do
    cond do
      bytes >= 1_073_741_824 -> "#{Float.round(bytes / 1_073_741_824, 2)} GB"
      bytes >= 1_048_576 -> "#{Float.round(bytes / 1_048_576, 2)} MB"
      bytes >= 1_024 -> "#{Float.round(bytes / 1_024, 2)} KB"
      true -> "#{bytes} B"
    end
  end

  defp format_bytes(_), do: "Unknown"
end

File: src/discord/formatter.ts
import type { DiscordEmbed, NotificationMessage } from "../types/index.js";

const COLORS = {
	green: 0x57f287,
	red: 0xed4245,
	yellow: 0xfee75c,
	purple: 0x9b59b6,
	orange: 0xe67e22,
	blue: 0x3498db,
} as const;

function formatBytes(bytes: number): string {
	if (bytes === 0) return "0 B";
	const k = 1024;
	const sizes = ["B", "KB", "MB", "GB", "TB"];
	const i = Math.floor(Math.log(bytes) / Math.log(k));
	return `${Number.parseFloat((bytes / k ** i).toFixed(2))} ${sizes[i]}`;
}

function truncate(str: string, maxLength: number): string {
	return str.length > maxLength ? `${str.slice(0, maxLength - 3)}...` : str;
}

export function formatEmbed(routingKey: string, payload: NotificationMessage): DiscordEmbed {
	if (routingKey === "ci.success") {
		const msg = payload as { repository?: string; status?: string; run_url?: string };
		return {
			title: "CI Build Succeeded",
			color: COLORS.green,
			url: msg.run_url,
			fields: [
				{ name: "Repository", value: msg.repository ?? "unknown", inline: true },
				{ name: "Status", value: msg.status ?? "complete", inline: true },
				{ name: "Result", value: "success", inline: true },
			],
		};
	}

	if (routingKey === "ci.failure") {
		const msg = payload as { repository?: string; status?: string; run_url?: string };
		return {
			title: "CI Build Failed",
			color: COLORS.red,
			url: msg.run_url,
			fields: [
				{ name: "Repository", value: msg.repository ?? "unknown", inline: true },
				{ name: "Status", value: msg.status ?? "complete", inline: true },
				{ name: "Result", value: "failure", inline: true },
			],
		};
	}

	if (routingKey === "pr.opened") {
		const msg = payload as {
			pr_number?: number;
			pr_title?: string;
			author?: string;
			pr_url?: string;
		};
		return {
			title: "Pull Request Opened",
			color: COLORS.yellow,
			url: msg.pr_url,
			fields: [
				{ name: "PR", value: `#${msg.pr_number ?? "?"}`, inline: true },
				{ name: "Title", value: truncate(msg.pr_title ?? "untitled", 100), inline: true },
				{ name: "Author", value: msg.author ?? "unknown", inline: true },
			],
		};
	}

	if (routingKey === "pr.merged") {
		const msg = payload as {
			pr_number?: number;
			pr_title?: string;
			author?: string;
			pr_url?: string;
		};
		return {
			title: "Pull Request Merged",
			color: COLORS.purple,
			url: msg.pr_url,
			fields: [
				{ name: "PR", value: `#${msg.pr_number ?? "?"}`, inline: true },
				{ name: "Title", value: truncate(msg.pr_title ?? "untitled", 100), inline: true },
				{ name: "Author", value: msg.author ?? "unknown", inline: true },
			],
		};
	}

	if (routingKey === "pr.closed") {
		const msg = payload as {
			pr_number?: number;
			pr_title?: string;
			author?: string;
			pr_url?: string;
		};
		return {
			title: "Pull Request Closed",
			color: COLORS.red,
			url: msg.pr_url,
			fields: [
				{ name: "PR", value: `#${msg.pr_number ?? "?"}`, inline: true },
				{ name: "Title", value: truncate(msg.pr_title ?? "untitled", 100), inline: true },
				{ name: "Author", value: msg.author ?? "unknown", inline: true },
			],
		};
	}

	if (routingKey === "downloads.complete") {
		const msg = payload as { name?: string; size?: number; category?: string; savePath?: string };
		return {
			title: "Download Complete",
			color: COLORS.green,
			fields: [
				{ name: "Name", value: truncate(msg.name ?? "unknown", 100), inline: false },
				{ name: "Size", value: formatBytes(msg.size ?? 0), inline: true },
				{ name: "Category", value: msg.category ?? "unknown", inline: true },
			],
		};
	}

	if (routingKey === "downloads.removed") {
		const msg = payload as { name?: string; size?: number; category?: string; savePath?: string };
		return {
			title: "Download Removed",
			color: COLORS.orange,
			fields: [
				{ name: "Name", value: truncate(msg.name ?? "unknown", 100), inline: false },
				{ name: "Size", value: formatBytes(msg.size ?? 0), inline: true },
				{ name: "Category", value: msg.category ?? "unknown", inline: true },
			],
		};
	}

	if (routingKey === "deploy.success") {
		const msg = payload as { repository?: string; status?: string; duration?: number };
		return {
			title: "Deployment Succeeded",
			color: COLORS.green,
			fields: [
				{ name: "Repository", value: msg.repository ?? "unknown", inline: true },
				{ name: "Status", value: msg.status ?? "complete", inline: true },
				...(msg.duration ? [{ name: "Duration", value: `${msg.duration}s`, inline: true }] : []),
			],
		};
	}

	if (routingKey === "deploy.failure") {
		const msg = payload as { repository?: string; status?: string; duration?: number };
		return {
			title: "Deployment Failed",
			color: COLORS.red,
			fields: [
				{ name: "Repository", value: msg.repository ?? "unknown", inline: true },
				{ name: "Status", value: msg.status ?? "failed", inline: true },
				...(msg.duration ? [{ name: "Duration", value: `${msg.duration}s`, inline: true }] : []),
			],
		};
	}

	if (routingKey === "polling.failure") {
		const msg = payload as { service?: string; error?: string };
		return {
			title: "Service Polling Failed",
			color: COLORS.red,
			fields: [
				{ name: "Service", value: msg.service ?? "unknown", inline: true },
				{ name: "Error", value: truncate(msg.error ?? "unknown error", 200), inline: false },
			],
		};
	}

	if (routingKey.startsWith("dlq.") || routingKey.startsWith("notifications.dlq.")) {
		const msg = payload as {
			service?: string;
			queue?: string;
			error?: string;
			retryCount?: number;
			originalMessage?: string;
		};
		return {
			title: "Dead Letter Queue Alert",
			color: COLORS.red,
			fields: [
				{ name: "Service", value: msg.service ?? "unknown", inline: true },
				{ name: "Queue", value: msg.queue ?? "unknown", inline: true },
				{ name: "Retry Count", value: String(msg.retryCount ?? 0), inline: true },
				{ name: "Error", value: truncate(msg.error ?? "unknown error", 200), inline: false },
			],
		};
	}

	if (routingKey === "warn.diskspace") {
		const msg = payload as {
			volume?: string;
			mount_point?: string;
			used_percent?: number;
			threshold?: string;
			total_bytes?: number;
			used_bytes?: number;
			free_bytes?: number;
		};
		return {
			title: "Disk Space Warning",
			color: COLORS.yellow,
			fields: [
				{ name: "Volume", value: msg.volume ?? "unknown", inline: true },
				{ name: "Usage", value: `${msg.used_percent ?? 0}%`, inline: true },
				{ name: "Free", value: formatBytes(msg.free_bytes ?? 0), inline: true },
			],
		};
	}

	if (routingKey === "error.diskspace") {
		const msg = payload as {
			volume?: string;
			mount_point?: string;
			used_percent?: number;
			threshold?: string;
			total_bytes?: number;
			used_bytes?: number;
			free_bytes?: number;
		};
		return {
			title: "Disk Space Critical",
			color: COLORS.orange,
			fields: [
				{ name: "Volume", value: msg.volume ?? "unknown", inline: true },
				{ name: "Usage", value: `${msg.used_percent ?? 0}%`, inline: true },
				{ name: "Free", value: formatBytes(msg.free_bytes ?? 0), inline: true },
			],
		};
	}

	if (routingKey === "critical.diskspace") {
		const msg = payload as {
			volume?: string;
			mount_point?: string;
			used_percent?: number;
			threshold?: string;
			total_bytes?: number;
			used_bytes?: number;
			free_bytes?: number;
		};
		return {
			title: "Disk Space Emergency",
			color: COLORS.red,
			fields: [
				{ name: "Volume", value: msg.volume ?? "unknown", inline: true },
				{ name: "Usage", value: `${msg.used_percent ?? 0}%`, inline: true },
				{ name: "Free", value: formatBytes(msg.free_bytes ?? 0), inline: true },
			],
		};
	}

	if (routingKey === "info.diskspace") {
		const msg = payload as {
			volume?: string;
			mount_point?: string;
			used_percent?: number;
			previous_threshold?: string;
			volumes?: Array<{
				volume: string;
				mount_point: string;
				used_percent: number;
				status: string;
			}>;
		};

		if (msg.previous_threshold) {
			return {
				title: "Disk Space Recovery",
				color: COLORS.green,
				fields: [
					{ name: "Volume", value: msg.volume ?? "unknown", inline: true },
					{ name: "Usage", value: `${msg.used_percent ?? 0}%`, inline: true },
					{ name: "Recovered From", value: msg.previous_threshold, inline: true },
				],
			};
		}

		if (msg.volumes && msg.volumes.length > 0) {
			const volumeList = msg.volumes
				.map((v) => `${v.volume}: ${v.used_percent}% (${v.status})`)
				.join("\n");
			return {
				title: "Disk Space Status",
				color: COLORS.blue,
				description: truncate(volumeList, 1000),
			};
		}

		return {
			title: "Disk Space Status",
			color: COLORS.blue,
			fields: [
				{ name: "Volume", value: msg.volume ?? "unknown", inline: true },
				{ name: "Usage", value: `${msg.used_percent ?? 0}%`, inline: true },
			],
		};
	}

	return {
		title: `Notification: ${routingKey}`,
		description: truncate(JSON.stringify(payload, null, 2), 1000),
		color: COLORS.blue,
	};
}
